core/edr-core/src/main/java/org/eclipse/tractusx/edc/edr/core/lock/InMemoryEdrLock.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2024 Bayerische Motoren Werke Aktiengesellschaft
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Apache License, Version 2.0 which is available at
+ * https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.eclipse.tractusx.edc.edr.core.lock;
+
+import org.eclipse.edc.edr.spi.store.EndpointDataReferenceEntryIndex;
+import org.eclipse.edc.spi.result.StoreResult;
+import org.eclipse.edc.spi.types.domain.DataAddress;
+import org.eclipse.edc.transaction.spi.TransactionContext;
+import org.eclipse.tractusx.edc.edr.spi.index.lock.EndpointDataReferenceLock;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+
+public class InMemoryEdrLock implements EndpointDataReferenceLock {
+
+    private static final ReentrantReadWriteLock LOCK = new ReentrantReadWriteLock();
+
+    private final EndpointDataReferenceEntryIndex entryIndex;
+    private final TransactionContext transactionContext;
+    private final Map<String, ReentrantReadWriteLock> lockedEdrs = new ConcurrentHashMap<>();
+
+    public InMemoryEdrLock(EndpointDataReferenceEntryIndex entryIndex, TransactionContext transactionContext) {
+        this.entryIndex = entryIndex;
+        this.transactionContext = transactionContext;
+    }
+
+    @Override
+    public StoreResult<Boolean> acquireLock(String edrId, DataAddress edr) {
+        LOCK.writeLock().lock();
+        try {
+            var rowLock = lockedEdrs.computeIfAbsent(edrId, k -> new ReentrantReadWriteLock());
+
+            LOCK.writeLock().unlock(); // unlocks global lock before waiting
+            rowLock.writeLock().lock();
+            LOCK.writeLock().lock();
+            // inner try loop for the row-level lock
+            var edrEntry = transactionContext.execute(() -> entryIndex.findById(edrId));
+
+            return StoreResult.success(isExpired(edr, edrEntry));
+        } finally {
+            LOCK.writeLock().unlock();
+        }
+    }
+
+
+    @Override
+    public StoreResult<Void> releaseLock(String edrId) {
+        LOCK.writeLock().lock();
+        try {
+            var reentrantReadWriteLock = lockedEdrs.get(edrId);
+            if (reentrantReadWriteLock == null) {
+                return StoreResult.generalError("Could not release row-lock because it does not exist");
```

Comment:
not sure this is really an error.

Reply:
I don't have a strong opinion on this either. If we are to handle the result from the call in the `EdrServiceImpl`, my guess was that we could at least be informed that some thread is trying to release the lock when it does already exist...

Reply:
If we handle it with a success, then it has no point in handling the result of it at all. The method doesn't have any return case that has to be handled...

Reply:
[64c76a9](https://github.com/eclipse-tractusx/tractusx-edc/pull/1633/commits/64c76a920083f1bb113088b65264520b2d909fb3)